tests/integration/high_availability/test_replication.py

```diff
@@ -255,8 +258,161 @@ async def test_no_replication_across_clusters(ops_test: OpsTest, continuous_writ
         assert database_name not in output
 
     # remove another mysql application cluster
-    await scale_application(ops_test, another_mysql_application_name, 0)
+    await scale_application(ops_test, another_mysql_application_name, 0, wait=False)
     await ops_test.model.remove_application(
         another_mysql_application_name,
-        block_until_done=True,
+        block_until_done=False,
+    )
+
+    # clean up inserted data, and created tables + databases
+    clean_up_database_and_table_sql = [
+        f"DROP TABLE IF EXISTS `{database_name}`.`{table_name}`",
+        f"DROP DATABASE IF EXISTS `{database_name}`",
+    ]
+    execute_queries_on_unit(
+        mysql_primary_address,
+        server_config_credentials["username"],
+        server_config_credentials["password"],
+        clean_up_database_and_table_sql,
+        commit=True,
+    )
+
+
+@pytest.mark.order(2)
+@pytest.mark.abort_on_fail
+@pytest.mark.replication_tests
+async def test_preserves_data_on_delete(ops_test: OpsTest) -> None:
+    """Test to ensure that data is preserved when a unit is scaled up and then down.
+
+    Ensures that there are no running continuous writes as the extra data in the
+    database makes scaling up slower.
+    """
+    mysql_application_name, _ = await high_availability_test_setup(ops_test)
+
+    # assert that there are 3 units in the mysql cluster
+    assert len(ops_test.model.applications[mysql_application_name].units) == 3
+
+    mysql_unit = ops_test.model.applications[mysql_application_name].units[0]
+    primary = await get_primary_unit(ops_test, mysql_unit, mysql_application_name)
+    primary_address = await get_unit_address(ops_test, primary.name)
+
+    # insert a value before scale up, and ensure that the value exists in all units
+    database_name, table_name = "test-preserves-data-on-delete", "data"
+    value_before_scale_up = generate_random_string(255)
+    server_config_credentials = await get_server_config_credentials(primary)
+    insert_before_scale_up_sql = [
+        f"CREATE DATABASE IF NOT EXISTS `{database_name}`",
+        f"CREATE TABLE IF NOT EXISTS `{database_name}`.`{table_name}` (id varchar(255), primary key(id))",
+        f"INSERT INTO `{database_name}`.`{table_name}` (id) VALUES ('{value_before_scale_up}')",
+    ]
+
+    await execute_queries_on_unit(
+        primary_address,
+        server_config_credentials["username"],
+        server_config_credentials["password"],
+        insert_before_scale_up_sql,
+        commit=True,
+    )
+
+    select_value_before_scale_up_sql = [
+        f"SELECT id FROM `{database_name}`.`{table_name}` WHERE id = '{value_before_scale_up}'",
+    ]
+
+    for attempt in Retrying(stop=stop_after_delay(10), wait=wait_fixed(2)):
```

Comment:
why the need for retrying in here, shouldn't the cluster be settled?

Reply:
Retrying for any replication lag that may be present